### PR TITLE
fix for infinite loop when time range is less than 20 micros

### PR DESCRIPTION
**Pull request by bradNASA**
* **Tickets addressed:** none
* **Review:** By commit 
* **Merge strategy:** Merge 

## Description
Infinite loop when searching in a time range of micros less than the max iterations where no solutions exist

## Verification
ran unit tests, and it fixed this bug occurring on another branch

## Documentation
none

## Future work
add a test case that searches within a <20 microsecond window

### CODE REVIEW

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java [hunk 1]
```diff
@@ -141,8 +144,8 @@ private IteratingResult nextValueAt(
           }
         }
         cur = chooseRandomX(min, max);
-        //if min == max, another call to random will have no effect and thus we should exit
-      } while(i < maxIteration && !min.isEqualTo(max));
+        //if all timepoints have been visited or min == max, another call to random will have no effect and thus we should exit
+      } while(i < maxIters);
```
Comment (adrienmaillard) [inline scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: why removing the second `!min.isEqualTo(max)` ?

Comment (adrienmaillard) [inline scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: I don't see how this leads to an infinite loop in the case you describe ? Can you elaborate either on the PR description or in an answer to that comment ?

**Review by adrienmaillard — COMMENTED** [review]
Some questions first

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java (hunk 1 above)
Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: if min == max, `numTimepoints` will be -1, so `maxIters` will be -1

Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: I came across a run that called this method with min = 4 micros and max = 6 micros, so only 4 or 5 would be randomly encountered.  At some point, the `if` statement always fails, and `i` is never incremented, so the loop never ends.

Comment (adrienmaillard) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: can you then remove the "or min == max" part of the comment ?

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java [hunk 2]
```diff
@@ -127,6 +127,9 @@ private IteratingResult nextValueAt(
         final int maxIteration)
     throws ExceededMaxIterationException, SchedulingInterruptedException
     {
+      // the number of possible values may be less than the number of iterations, so stop after all have been visited.
+      long numTimepoints = max.in(Duration.MICROSECONDS) - min.in(Duration.MICROSECONDS) - 1;
```
Comment (adrienmaillard) [inline scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java]: this is a numeric overflow if `max` is `Long.MAX_VALUE` and `min` is `Long.MIN_VALUE`.

Comment (adrienmaillard) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java]: maybe use `BigInteger` ?

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java (hunk 1 above)
Comment (adrienmaillard) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: OK! it sounds like the increment of `i` is at fault here but that's a reasonable workaround.

Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: removed

Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java:148]: `i` was only incremented for values that weren't already visited.  So, I don't think always incrementing `i` is a good fix.  It think this fix is appropriate.

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java (hunk 2 above)
Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java]: I pasted in some code for handling overflow.

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java [hunk 3]
```diff
@@ -127,25 +127,56 @@ private IteratingResult nextValueAt(
...
-      } while(i < maxIteration && !min.isEqualTo(max));
+        //if all timepoints have been visited, another call to random will have no effect and thus we should exit
+      } while(i < maxIters);
       throw new ExceededMaxIterationException();
     }
 
+    /**
+     * Addition of longs with safety from overflow
+     */
+    private static long plus( long rd1, long rd2 ) {
```
Comment (adrienmaillard) [inline scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java]: can you final rd1 and rd2 ? same in the minus.

Comment (bradNASA) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java]: "final" modifiers added to plus() and minus()

**Review by dandelany — APPROVED** [review]
Thanks Brad - all issues above resolved & tests passing 👍 I'll approve and merge this on @adrienmaillard 's behalf since he's out this week.

**Pull request merged**